src/jupyterlabcontroller/services/fileserver.py

```diff
@@ -0,0 +1,422 @@
+        self._logger = logger
+        self.config = config
+        self.k8s_client = k8s_client
+        self._tasks: Set[asyncio.Task] = set()
+
+    async def create_fileserver_if_needed(self, user: UserInfo) -> bool:
+        """If the user doesn't have a fileserver, create it.  If the user
+        already has a fileserver, just return.
+
+        This gets called by the handler when a user comes in through the
+        /file ingress.
+        """
+        username = user.username
+        self._logger.info(f"Fileserver requested for {username}")
+        if not self.user_map.get(username):
+            try:
+                if not await self.create_fileserver(user):
+                    return False
+            except KubernetesError:
+                self._logger.error(
+                    f"Fileserver creation for {username} failed: deleting "
+                    + "fileserver objects."
+                )
+                await self.delete_fileserver(username)
+                raise
+            self.user_map.set(username)
+        return True
+
+    async def create_fileserver(self, user: UserInfo) -> bool:
+        """Create a fileserver for the given user.  Wait for it to be
+        operational.  If we can't build it, return False.
+        """
+        username = user.username
+        self._logger.info(f"Creating new fileserver for {username}")
+        namespace = self.fs_namespace
+        if not await self.k8s_client.check_namespace(namespace):
+            self._logger.warning("No fileserver namespace '{namespace}'")
+            return False
+        job = self.build_fileserver_job(user)
+        gf_ingress = self.build_fileserver_ingress(user.username)
+        service = self.build_fileserver_service(user.username)
+        # We put this all in k8s_client so it can manage its own lock
+        return await self.k8s_client.create_fileserver(
+            username=username,
+            namespace=namespace,
+            job=job,
+            service=service,
+            gf_ingress=gf_ingress,
+        )
+
+    def build_fileserver_metadata(self, username: str) -> V1ObjectMeta:
+        """Construct metadata for the user's fileserver objects.
+
+        Parameters
+        ----------
+        username
+            User name
+
+        Returns
+        -------
+        V1ObjectMeta
+            Kubernetes metadata specification for that user's fileserver
+            objects.
+        """
+        obj_name = f"{username}-fs"
+        return V1ObjectMeta(
+            name=obj_name,
+            namespace=self.fs_namespace,
+            labels={
+                "argocd.argoproj.io/instance": "fileservers",
+                "lsst.io/category": "fileserver",
+                "lsst.io/user": username,
+            },
+            annotations={
+                "argocd.argoproj.io/compare-options": "IgnoreExtraneous",
+                "argocd.argoproj.io/sync-options": "Prune=false",
+            },
+        )
```

Comment:
This duplicates `_standard_metadata` in `K8sStorageClient` except for a different Argo CD project and the `lsst.io` annotations. I think those functions should be merged at an appropriate architectural level. The category and user metadata seems helpful, or at least harmless, for lab objects as well.

Reply:
I can do that.  I didn't want to start exporting a bunch more stuff from the storage client, but, OK.

Reply:
I've done this although I feel like ultimately the metadata builder belongs in the Builder object, not the storage client.